### PR TITLE
Specialize Serialization for OrderedDict and defaultdict

### PR DESCRIPTION
**Pull request by pcmoritz**
Before this gets merged into the mainline, need to do:

- [x] Investigate performance tradeoffs and optimize code
- [x] Support other dictionaries
- [x] Merge changes into arrow

### CODE REVIEW

**Comment by pcmoritz** [comment]
Performance investigation of the first naive version:

```python
import ray
import collections
ray.init()

d = collections.OrderedDict([("hello-" + str(i), i) for i in range(100000)])
%time x = ray.put(d)
# Wall time: 53.9 ms
%time y = ray.get(x)
# Wall time: 55.4 ms

import pickle
%time x = pickle.dumps(d)
# Wall time: 43.9 ms
%time y = pickle.loads(x)
# Wall time: 59.8 ms
```

File: python/ray/worker.py [hunk 1]
```diff
@@ -1040,6 +1040,17 @@ def array_custom_deserializer(serialized_obj):
         custom_serializer=array_custom_serializer,
         custom_deserializer=array_custom_deserializer)
 
+    def ordereddict_custom_serializer(obj):
+        return list(obj.keys()), list(obj.values())
```
Comment (richardliaw) [inline python/ray/worker.py]: why not 1 list of tuples? ie, `return list(obj.items())`

this way the deserializer just needs to do `return collections.OrderedDict(obj)`

Comment (pcmoritz) [reply python/ray/worker.py]: Yeah, that's a good point! I'm currently trying to get rid of the temporary copy altogether, let's see how that goes :)

**Comment by pcmoritz** [comment]
Getting rid of the extra copy (as in the latest commit):

```python
%time x = ray.put(d)
# Wall time: 56.8 ms
```

I'm a little surprised that there is no speedup!

And with Richard's solution
```python
In [5]: %time x = ray.put(d)
CPU times: user 70.3 ms, sys: 8.4 ms, total: 78.7 ms
Wall time: 77.1 ms

In [6]: %time y = ray.get(x)
CPU times: user 149 ms, sys: 833 µs, total: 150 ms
Wall time: 149 ms
```

Again surprising...

Anyways, we should then use the first solution ;)

**Comment by robertnishihara** [comment]
This should fix #512 and #1034.

**Review by robertnishihara — APPROVED** [review]
This looks good to me!

**Comment by mitar** [comment]
Thank you for quick response!

**Comment by robertnishihara** [comment]
You're welcome! Let us know if you run into any issues with the code.

**Pull request merged**